evaluate pass fail likelihoods

oindrillac opened the pull request:
In this notebook:

- I evaluate the trends of passing and failing tests and evaluate the likelihood of a test to pass or fail at a given point
- Instead of adding a function directly, I kept examples which describe the approach, create plots on different tests to get feedback on the method before proceeding to create a function for it
- I evaluate the approach on a larger dataset collected from the https://github.com/github/codeql repo and find a turning point **after which the test is 75% likely to fail** and at all points after then **it is never less than 60% likely to fail**. (These percentages can be tweaked). I then evaluate it on the test dataset and get an accuracy of 92.24%

**Some things to note**
- From some plots it is clear that there is no stopping point - meaning it never occurs that a test is more likely to fail than pass
- One can argue that a calculation of such a stopping point is skewed by the number of tests in each bucket (pass/fail), so I also added a section where I calculate a percent of percents which completely ignores the number of tests in each bucket. My intuition is that in the real scenario, this is not valid since as should also factor for the proportion of passing and failing tests. I'm open to your feedback

MichaelClifford commented:
:+1: Looks good @oindrillac

* I think it makes sense that there will be tests where there is no _optimal_ stopping point other than the max value of passed tests, (even if no failure in the dataset ever gets that far)

* I agree, I think the proportions matter.  Instead of calculating a percent of percent. Could you normalize the data so that each bucket has roughly the same number of passes and failures. And then get the % failure for each bucket? This seems like it would answer the question, for any Bucket at time T, here is the likelihood of failure.

oindrillac commented:
Thanks for reviewing @MichaelClifford!

Just so I am clear do you recommend normalizing the frequency counts within each time bucket?
I just tried that here and pushed it to this branch. Here, I am scaling the number of tests within each 30 second time range to be between 0 & 1  and as a result you can see that in the prefix sum (cumulative sum) column, the first value which refers to all the tests that passed or failed from time 0 is pretty similar for both. The osp here however is much lower than the previous attempt and the accuracy is pretty similar.
https://github.com/oindrillac/optimal-stopping-point-tool/blob/normalize-data/notebooks/pass_fail_likelihoods.ipynb

Is this what you were recommending? Or I am misunderstanding your suggestion

oindrillac commented:
Fyi - You need to scroll to the very bottom to look at the normalization

MichaelClifford commented:
@oindrillac yeah. I think that is right. 

Also, the passing values are all just 1-failing the values. So I don't think we need to plot them both. They are basically just 2 views of the same underlying distribution. 

So what we have here is: for every time interval _T_, we sum all the successes from _T_ to _N_ (the end if the dataset) divided by the total number of tests from _T_ to _N_. This will give us an answer to the question. Given it is now time _T_, what is the probability that my test will succeed should I allow it to continue.  :+1: That's what you had before, and I think is totally right. I think the normalization helps here, but I'm actually not sure if its strictly necessary as we are loosing data each time we step forwards in time anyways, right? The problem is necessarily skewed by a fewer number of examples at each time step, but the relative number of pass and fail tests shouldn't be skewed.  Let's revisit this and see how much it actually makes sense to do. :smile:    

Once we have our percent success for each time step, then we say, for some user defined threshold _E_, what is the soonest time that my probability of success is less than or equal to _E_ and never again larger than _E_.

And that values is what we consider to be our Optimal Stopping Point.     

I'm not saying anything new here I don't think, just reiterating what it looks like you're notebook is doing. Please correct me if I'm wrong. 

Otherwise, this looks right to me.

oindrillac commented:
Thank you for giving this a detailed look @MichaelClifford!

> So what we have here is: for every time interval _T_, we sum all the successes from _T_ to _N_ (the end if the dataset) divided by the total number of tests from _T_ to _N_. 

Yes but here I am dividing the sum of all the successes from _T_ to _N_ by  by the total number of tests from _T0_ to _N_ or the total number of tests in that set. 

> Given it is now time _T_, what is the probability that my test will succeed should I allow it to continue. 👍 

Yes, that is exactly what I am doing

> but I'm actually not sure if its strictly necessary as we are loosing data each time we step forwards in time anyways, right? 

Right, we can evaluate this a little bit more whether normalizing helps

> Once we have our percent success for each time step, then we say, for some user defined threshold _E_, what is the soonest time that my probability of success is less than or equal to _E_ and never again larger than _E_.

Slight correction - We are optimizing for 2 variables here. Although we could also do what you mentioned.
What I am doing here is for some user defined threshold _E_ and _M_, what is the soonest time that my probability of success is less than or equal to _M_ and never again larger than _E_.
for example if E=25 and M=40, we get the point at which the probability of passing is only 25% and after which the probability of passing is never more than 40%

I want to evaluate this for a few more tests and see how normalization impacts the results and use some more metrics beyond accuracy to evaluate how this is doing.

oindrillac commented:
cc: @erikerlandson

oindrillac commented:
Made the following changes from the last time you had a look

- Added more metrics for evaluation
- Added normalization and compared the results
- Running this on 3 larger tests now where we compare the results with and without normalizing

MichaelClifford reviewed (approved):
+1 LGTM

aakankshaduggal reviewed (commented):
Thanks @oindrillac, looks great!
Can you please re-run the notebooks so the cells are numbered in the right execution order?
Otherwise /lgtm

oindrillac commented:
Thanks! 
I consolidated everything into one notebook, documented the result (no impact) of normalizing that we are seeing and re-ran all the cells.
